gcloud-java-core/src/main/java/com/google/gcloud/ServiceOptions.java

```diff
@@ -479,6 +487,20 @@ protected static String getAppEngineProjectId() {
     }
   }
 
+  protected static String serviceAccountProjectId() {
+    String project = null;
+    String credentialsPath = System.getenv("GOOGLE_APPLICATION_CREDENTIALS");
+    if(credentialsPath != null) {
+      try (InputStream credentialsStream = new FileInputStream(credentialsPath)) {
```

Comment:
This should never be executed on AE as the `GOOGLE_APPLICATION_CREDENTIALS` is not set there, right? BTW you are right we better keep the same order that we have for credentials.